anxcloud/resource_dns_record.go

```diff
@@ -0,0 +1,225 @@
+package anxcloud
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+	clouddnsv1 "go.anx.io/go-anxcloud/pkg/apis/clouddns/v1"
+	"go.anx.io/go-anxcloud/pkg/utils/object/compare"
+)
+
+var syncDNSRecordOps sync.Mutex
+var dnsRecordSkipMutexLock = providerContextKey("dns-record-skip-mutex-lock")
+
+func resourceDNSRecord() *schema.Resource {
+	return &schema.Resource{
+		Description:   "This resource allows you to create DNS records for a specified zone. Records of type TXT might be unstable for now.",
+		CreateContext: resourceDNSRecordCreate,
+		ReadContext:   resourceDNSRecordRead,
+		UpdateContext: resourceDNSRecordUpdate,
+		DeleteContext: resourceDNSRecordDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(time.Minute),
+			Read:   schema.DefaultTimeout(time.Minute),
+			Update: schema.DefaultTimeout(time.Minute),
+			Delete: schema.DefaultTimeout(time.Minute),
+		},
+		Schema: schemaDNSRecord(),
+	}
+}
+
+func resourceDNSRecordCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	defer lockRecordContextIfNeeded(ctx)()
+
+	a := m.(providerContext).api
+
+	r := dnsRecordFromResourceData(d)
+
+	// try to import
+	if ret, err := findDNSRecord(ctx, a, r); api.IgnoreNotFound(err) != nil {
+		return diag.FromErr(err)
+	} else if err == nil {
+		// DNS Record found
+		d.SetId(ret.Identifier)
+		return resourceDNSRecordRead(ctx, d, m)
+	}
+
+	// not found -> create new zone
+
+	if err := a.Create(ctx, &r); err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(r.Identifier)
```

Comment:
You are using the record identifier as terraform resource identifier. How does this work with the identifier changing, especially when the change is not triggered by this record resource changing but any other in that zone? Is Terraform fine with the ID randomly changing?

I guess since the ID of the terraform record resource is never used by the provider, this is fine as it only changes for Create and Update in terraform?

Reply:
correct. as long as the identifier is unique this should be fine